pkg/telemetryquerylanguage/tql/functions.go

```diff
@@ -44,33 +44,53 @@ func NewFunctionCall(inv Invocation, functions map[string]interface{}, pathParse
 	return nil, fmt.Errorf("undefined function %v", inv.Function)
 }
 
-func buildArgs(inv Invocation, fType reflect.Type, functions map[string]interface{}, pathParser PathExpressionParser, enumParser EnumParser) ([]reflect.Value, error) {
+func (p *Parser) buildArgs(inv Invocation, fType reflect.Type) ([]reflect.Value, error) {
 	args := make([]reflect.Value, 0)
+	// Some function arguments may be intended to take values from the calling processor
+	// instead of being passed by the caller of the TQL function, so we have to keep
+	// track of the index of the argument passed within the DSL.
+	// e.g. zap.Logger, which is provided by the processor to the TQL Parser struct.
+	DSLArgumentIndex := 0
 	for i := 0; i < fType.NumIn(); i++ {
 		argType := fType.In(i)
 
-		if argType.Kind() == reflect.Slice {
-			err := buildSliceArg(inv, argType, i, &args, functions, pathParser, enumParser)
+		switch argType.Kind() {
+		case reflect.Slice:
+			err := p.buildSliceArg(inv, argType, i, &args)
 			if err != nil {
 				return nil, err
 			}
-		} else {
-			if i >= len(inv.Arguments) {
+			// Slice arguments must be the final argument in an invocation.
+			return args, nil
+		case reflect.Pointer:
+			switch argType.Elem().String() {
+			case "zap.Logger":
```

Comment:
What if instead of a specific logging framework type we made a new function type, something like `type WriteLog func(string) (error)` that would wrap the logger?  Any functions that need to log would add this type to the parameters list and the Parser struct would have a field of this type instead of specifically zap.logger.  When calling `NewParser` the function could be passed in.  When a function needs to write a log it would pass the string to `WriteLog`.

This allows users to choose their own logging framework and for us to use zap.logger.  And then in the processor's Validate function we might be able to get away with passing in a No-Op instance of WriteLog instead of creating a new Logger instance.

Reply:
I agree with this in general -- we definitely shouldn't depend on any particular third party Logger type. 

I do, however, think that (since we are, in fact, talking about Telemetry here) we should support a structured Log interface, where an arbitrary collection of keys and values can be added to a log. Perhaps we could just boil it down to something that can be implemented as wrappers around nearly any logger, and it's also easy to mock for testing.

```go
type Logger interface {
      WithFields(field map[string]any) Logger
      Print(msg string)
}
```

Reply:
@kentquirk I like the interface idea, but can you expand more on the `WithFields` concept?

Reply:
Structured logs -- the log message can include a collection of key-value pairs, which means that the logs can be output in a structured format (JSON, say), making them queryable in a variety of logging targets.

From the zap docs: 
> Like log15 and go-kit, the SugaredLogger's structured logging APIs are loosely typed and accept a variadic number of key-value pairs.

Reply:
Ok I see what you mean.  Yes, we want to support zap's ability to pass in key value pairs, and my wrap idea wouldn't have supported that.  Are you picturing using the logging interface would look like:

```Go
func AFunction(logger Logger) (tql.ExprFunc, error) {
	logger.WithFields(...).Print("a message")
}
```

Where the fields passed to `WithFields` are then inserted into every string passed to `Print`?

Reply:
I would also much prefer to use an interface instead of directly depending on a concrete logging type, but since all other components use a Zap logger provided by the Collector, it's not immediately clear to me why a user would opt to use two loggers while running the Collector. Would it make sense to look ahead at having the Collector also allow users to provide a logging framework? Alternatively, do we expect that the TQL will be used outside the Collector?

Regarding the interface itself, I think it would also make sense to allow for specifying a log level when printing a message. The simplest way I can think of to do this would be to declare a handful of `LogLevel` integer constants, then users could adapt each value to their logging framework. User-written TQL functions could then also use custom log levels by using integers not defined in the constants.

Reply:
Yes, @TylerHelmuth, the way you've used it is the one-line way to do it. The thing I was talking about is that you can be given a logger, add some fields to it, and then pass the result to something else that's also expecting a logger. Then everything that second logger is doing has those fields added.

@evan-bradley I think the hope is that TQL is available for use elsewhere (although the Collector is the primary target). I also just think it's best practice not to depend on an interface defined by a third party. It's also easier to mock (I did see that zap has a parallel in zaptest, but that makes the problem worse IMO.)

Regarding log levels, I'm personally of mixed minds about it. With structured logging, I almost never use them, but I know some people are addicted to them. We could either pass in an enum as you suggest (which is smaller), or expand the interface to include, say, Info(), Debug(), Warn(), and Error() methods (which seems to be more idiomatic across logging libraries). But honestly, I'm not sure that TQL needs that much subtlety. Feels to me like if we do it at all, Info() and Error() are probably all we need.

Reply:
> he hope is that TQL is available for use elsewhere
Correct.  Since this lives in `pkg`, it can be used outside the collector.

Let's move forward with the interface idea as @kentquirk has described it.  I'm ok starting out with only Info and Error to make it easier to provide the logging struct to NewParser.